File: serverless-functions/report-generator/handler.go
package handler

import (
	"fmt"
	"encoding/json"
	"log"
)

// SubmissionRequest represents the submission data
type SubmissionRequest struct {
	SubmissionID int64  `json:"submissionId"`
	QuizID       int64  `json:"quizId"`
	UserID       int64  `json:"userId"`
	Score        int    `json:"score"`
	MaxScore     int    `json:"maxScore"`
	Title        string `json:"title"`
	SubmittedAt  string `json:"submittedAt"`
}

// ReportResponse represents the generated report
type ReportResponse struct {
	Status   string `json:"status"`
	ReportID string `json:"reportId"`
	Message  string `json:"message"`
	PDF      string `json:"pdfUrl"`
}

// Handle generates a PDF report for a quiz submission
func Handle(w http.ResponseWriter, r *http.Request) {
	if r.Method == "POST" {
		var submission SubmissionRequest
		
		// Parse request body
		err := json.NewDecoder(r.Body).Decode(&submission)
		if err != nil {
			http.Error(w, "Invalid request body", http.StatusBadRequest)
			return
		}
		
		log.Printf("📄 Generating report for submission %d (Quiz: %s)", submission.SubmissionID, submission.Title)
		
		// Calculate percentage
		percentage := 0
		if submission.MaxScore > 0 {
			percentage = (submission.Score * 100) / submission.MaxScore
		}
		
		// Generate report (in production, this would create actual PDF)
		reportID := fmt.Sprintf("report-%d-%d", submission.SubmissionID, submission.QuizID)
		
		report := ReportResponse{
			Status:   "success",
			ReportID: reportID,
			Message:  fmt.Sprintf("Report generated for %s. Score: %d/%d (%d%%)", submission.Title, submission.Score, submission.MaxScore, percentage),
			PDF:      fmt.Sprintf("https://reports.example.com/%s.pdf", reportID),
		}
		
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(report)
		
		log.Printf("✅ Report %s generated successfully", reportID)
	} else {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

File: serverless-functions/report-generator/handler.py
import json
import logging
from datetime import datetime

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def handle(req):
    """
    Generate a PDF report for a quiz submission.
    Receives submission data and generates a report.
    
    Expected input:
    {
        "submissionId": 123,
        "quizId": 456,
        "userId": 789,
        "score": 15,
        "maxScore": 20,
        "title": "Math Quiz",
        "submittedAt": "2026-01-08T10:30:00"
    }
    """
    try:
        # Parse request
        body = json.loads(req)
        
        submission_id = body.get("submissionId")
        quiz_id = body.get("quizId")
        user_id = body.get("userId")
        score = body.get("score", 0)
        max_score = body.get("maxScore", 100)
        title = body.get("title", "Quiz Report")
        submitted_at = body.get("submittedAt", "")
        
        # Calculate percentage
        percentage = (score * 100) // max_score if max_score > 0 else 0
        
        # Generate report metadata
        report_id = f"report-{submission_id}-{quiz_id}"
        timestamp = datetime.now().isoformat()
        
        logger.info(f"📄 Generating report {report_id} for submission {submission_id}")
        
        # In production, this would generate an actual PDF file
        # For now, we'll return report metadata
        report = {
            "status": "success",
            "reportId": report_id,
            "submissionId": submission_id,
            "quizId": quiz_id,
            "userId": user_id,
            "title": title,
            "score": f"{score}/{max_score}",
            "percentage": f"{percentage}%",
            "generatedAt": timestamp,
            "pdfUrl": f"https://reports.example.com/{report_id}.pdf",
            "message": f"Report generated for '{title}'. Score: {score}/{max_score} ({percentage}%)"
        }
        
        logger.info(f"✅ Report {report_id} generated successfully")
        
        return json.dumps(report)
        
    except json.JSONDecodeError:
        logger.error("Invalid JSON in request")
        error_response = {
            "status": "error",
            "message": "Invalid JSON in request body"
        }
        return json.dumps(error_response)
    except Exception as e:
        logger.error(f"❌ Error generating report: {str(e)}")
        error_response = {
            "status": "error",
            "message": f"Error generating report: {str(e)}"
        }
        return json.dumps(error_response)
